storage/storage.go

```diff
@@ -31,22 +32,211 @@ func NewPaginationOptions(ps int32, contToken string) PaginationOptions {
 	}
 }
 
-// TupleIterator is an iterator for Tuples. It is closed by explicitly calling Stop() or by calling Next() until it
-// returns an TupleIteratorDone error.
-type TupleIterator interface {
-	Next() (*openfgapb.Tuple, error)
-	// Stop will release any resources held by the iterator. It must be safe to be called multiple times.
+type Iterator[T any] interface {
+	Next() (T, error)
 	Stop()
 }
 
 // ObjectIterator is an iterator for Objects (type + id). It is closed by explicitly calling Stop() or by calling Next() until it
-// returns an ObjectIteratorDone error.
-type ObjectIterator interface {
-	Next() (string, error)
-	// Stop will release any resources held by the iterator. It must be safe to be called multiple times.
-	Stop()
+// returns an IteratorDone error.
+type ObjectIterator = Iterator[*openfgapb.Object]
+
+// TupleIterator is an iterator for Tuples. It is closed by explicitly calling Stop() or by calling Next() until it
+// returns an IteratorDone error.
+type TupleIterator = Iterator[*openfgapb.Tuple]
+
+// TupleKeyIterator is an iterator for TupleKeys. It is closed by explicitly calling Stop() or by calling Next() until it
+// returns an IteratorDone error.
+type TupleKeyIterator = Iterator[*openfgapb.TupleKey]
+
+// UniqueObjectIterator takes two ObjectIterators and yields an ObjectIterator that returns only distinct objects with the
+// duplicates removed.
+func UniqueObjectIterator(iter1, iter2 ObjectIterator) (ObjectIterator, error) {
+	objectMap := map[string]*openfgapb.Object{}
+
+	for {
+		val, err := iter1.Next()
+		if err != nil {
+			if errors.Is(err, IteratorDone) {
+				break
+			}
+			return nil, err
+		}
+
+		objectMap[tuple.ObjectKey(val)] = val
+	}
+
+	for {
+		val, err := iter2.Next()
+		if err != nil {
+			if errors.Is(err, IteratorDone) {
+				break
+			}
+			return nil, err
+		}
+
+		objectMap[tuple.ObjectKey(val)] = val
+	}
+
+	iter := &channelledStaticIterator[*openfgapb.Object]{
+		iteratorCh: make(chan *openfgapb.Object, len(objectMap)),
+		size:       uint64(len(objectMap)),
+	}
+	for _, object := range objectMap {
+		iter.iteratorCh <- object
+	}
+
+	return iter, nil
+}
+
+type combinedIterator[T any] struct {
+	iter1, iter2 Iterator[T]
+}
+
+func (c *combinedIterator[T]) Next() (T, error) {
+	val, err := c.iter1.Next()
+	if err != nil {
+		if !errors.Is(err, IteratorDone) {
+			return val, err
```

Comment:
```suggestion
			return nil, err
```

Reply:
Ditto the 2 below.

Reply:
The generic type `T` cannot take the value `nil`. So this is necessary.